### PR TITLE
[ACA-4728] fix file order in viewer

### PR DESCRIPTION
**Pull request by g-jaskowski**
**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [x] Tests for the changes have been added (for bug fixes / features)
> - [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

> - [x] Bugfix
> - [ ] Feature
> - [ ] Code style update (formatting, local variables)
> - [x] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:

**What is the current behaviour?** (You can also link to an open issue here)
Files browsed in viewer with previous/next buttons have different order than in document list.
https://alfresco.atlassian.net/browse/ACA-4728

**What is the new behaviour?**
Getting file order in viewer is set accordingly to sorting mode and unified between pages and viewer.
Last sorting key is saved and used to prevent datatable rows changing order after closing viewer.
Folder structure is changed to reduce code duplication.

**Does this PR introduce a breaking change?** (check one with "x")

> - [x] Yes
> - [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Review by MichalKinas — APPROVED** [review]

File: projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts:1]: @g-jaskowski @MichalKinas what do you think about creating single component instead of two different components (preview and viewer)? 
Both components look almost the same, they have even similar unit tests. It would simplify maintenance in the future because changes would be required only in single component instead of in two. 
This task is just bug fix so probably we would not like to do it here but maybe in separate task, what do you think? Would that be possible and good idea or better keep them as separate components?

**Review by AleksanderSklorz — APPROVED** [review]

**Comment by AleksanderSklorz** [comment]
@g-jaskowski before merging that PR please check sonarcloud issues

File: projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts
Comment (MichalKinas) [reply projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts:1]: I think it might not be that easy as preview has another custom use case, when client has a custom plugin e.g. starting a new process he can use preview component to check if file that he wants to attach to the process is correct. We would need to check that use case and investigate if those can be merged then

Comment (AleksanderSklorz) [reply projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts:1]: Thanks for response @MichalKinas. So should @g-jaskowski raise task to investigate that possibility and eventually merge these 2 components together if possible?

Comment (MichalKinas) [reply projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts:1]: Yes that's good idea to investigate that separately

Comment (AleksanderSklorz) [reply projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts:1]: great, then @g-jaskowski could you please raise task for this and put it to your team's board?

**Review by DenysVuika — APPROVED** [review]

**Comment by g-jaskowski** [comment]
I added `do not merge` label, as additional fixes are needed

**Comment by g-jaskowski** [comment]
@AleksanderSklorz @MichalKinas I tried to reduce code duplication with the refactor, it's still quite high, but can be lowered with some unit test optimization.
Also another small PR will be needed for ADW, but I wanted to make sure I took the right approach.

File: projects/aca-content/viewer/src/lib/components/preview/preview.component.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/components/preview/preview.component.ts:55]: is that used somewhere? I can't see usages so probably can be removed

File: projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts:185]: I would change a bit title because this test checks if getNodeInfo and navigate functions are not called. In reality it does not check if node is not displayed because there is no 'expect' which directly checks displaying of node

File: projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/components/viewer/viewer.component.spec.ts:225]: instead of mocking JSON.parse please try to prepare error object returned by getNodeInfo which gives you expected result in JSON.parse

File: projects/aca-content/viewer/src/lib/components/viewer/viewer.component.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/components/viewer/viewer.component.ts:87]: As you are doing changes here, could you remove that field because it seems unused?

File: projects/aca-content/viewer/src/lib/services/viewer.service.spec.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.spec.ts:29]: path can be just:  './viewer.service'

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:66]: source parameter seems missing in JSDoc above

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:98]: source parameter seems missing in JSDoc above

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:67]: can we have a type for this object and then please use this type as return type of this function

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:72]: this code can be a bit shorter: 
```ts
if (nodeId && folderId) {
      try {
        const ids = await this.getFileIds(source, folderId);
        const idx = ids.indexOf(nodeId);

        if (idx >= 0) {
          return {
            left: ids[idx - 1] || null,
            right: ids[idx + 1] || null
          };
        }
      } catch {}
}
return empty;

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:112]: This part: 
```ts
const entries = nodes.list.entries.map((obj) => obj.entry);
      if (isClient) {
        if (previousSortKey) {
          this.sort(entries, previousSortKey, previousSortDir);
        }
        this.sort(entries, sortKey, sortDirection);
      }
      return entries.map((entry) => entry.id);
```

looks similarly for every 'if'. Can you remove duplications? They are multiple ways to achieve that, for example it can be moved after all 'ifs' at the end of function so will be executed for every case

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:219]: missing return type

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:226]: can we use some type instead of any?

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:232]: same here, can we remove any?

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:245]: missing return type

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:249]: missing return type

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:257]: missing return type

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:70]: if allowed values for source are 

personal-files, libraries, favorites, shared, recent-files

then maybe we can limit type of source parameter to only these values instead of general string which allows every string?

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:100]: similarly like above - maybe we can limit type of source to only these allowed values instead of general string?

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:30]: type should start from upper case

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (g-jaskowski) [reply projects/aca-content/viewer/src/lib/services/viewer.service.ts]: @AleksanderSklorz I indiscriminately reverted to old param description. Now the description is corrected, so is the getFileIds function with 'wrong' source. Your solution would be cleaner, but the source is tied to pages, and due to libraries pages' logic I believe this solution to be sufficient.

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (g-jaskowski) [reply projects/aca-content/viewer/src/lib/services/viewer.service.ts:191]: Throughout whole ACA 'asc' is used, 'ASC' version occurred in one template and caused small bugs to happen with sorting indicator in column header. The direction is used in sorting: string[] property, so I left it without a type, as "...==='asc'"

File: projects/aca-content/viewer/src/lib/services/viewer.service.ts
Comment (AleksanderSklorz) [inline projects/aca-content/viewer/src/lib/services/viewer.service.ts:175]: No need to check if nodes is undefined because of multiple reasons: 
1. There is Intellij hint which says that it will give always false (because nodes will be never undefined): 
image

2. It can't be undefined because source can have following values: personal-files, libraries, favorites, shared and recent-files and you have 'if' for each of these values and in every 'if' you set nodes variable. 

3. You still refer to field of nodes variable (nodes.list) in 168th line so that means it is never undefined as you never get issue in that line that you refer to field of undefined. 

Hence please remove that condition which checks if nodes is undefined.

**Review by AleksanderSklorz — APPROVED** [review]

**Review by DenysVuika — APPROVED** [review]

**Pull request merged**
